torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -1111,7 +1083,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // Number of devices on this node.
   int localDeviceCount_{0};
 
-  ProcessGroupStatus pgStatus_;
+  std::shared_ptr<ProcessGroupStatus> pgStatus_ =
+      std::make_shared<ProcessGroupStatus>();
```

Comment:
this introduces possible race conditions that multiple threads are accessing the PgStatus. The dump threads (monitoring thread) and the write thread of watchdog threads. You will need to to introduce synchronization schemes such as locks.

Reply:
Hmmm, well. The race is here already even before we change it to a pointer? Before this change we do use it in the monitoring thread as well. You might argue that it was only used by log messages, but the race was indeed there already. We probably don't want to use a lock in monitoring thread and dump because what if watchdog thread holds the lock and it hangs. The dump is just best-effort so it's ok to dump stale data?

Reply:
A read/write lock could be optimal in this case. But that's true the stale data read is not a big problem here.

Reply:
what I can do here is to make it an atomic?

Reply:
We only dump when we detect we aren't making forward progress and are signaled to dump. 
If you're hung and not making forward progress, one could argue that new data is no longer being written to the buffer (i.e. there is no stale data)?
Is this a valid assumption?